webpack/assets/javascripts/react_app/components/EditableTextDisplay/EditableTextDisplay.js

```diff
@@ -0,0 +1,136 @@
+import React, { useState, useEffect } from 'react';
+import {
+  Spinner,
+  TextInput,
+  TextArea,
+  Text,
+  TextVariants,
+  Button,
+  Split,
+  SplitItem,
+} from '@patternfly/react-core';
+import { TimesIcon, CheckIcon, PencilAltIcon } from '@patternfly/react-icons';
+import PropTypes from 'prop-types';
+import { translate as __ } from '../../common/I18n';
+import './editableTextDisplay.scss';
+
+const EditableTextDisplay = ({ onEdit, value, textArea, attribute }) => {
+  // Tracks input box state
+  const [inputValue, setInputValue] = useState(value);
+  const [editing, setEditing] = useState(false);
+  const [submitting, setSubmitting] = useState(false);
+
+  // Setting didCancel to avoid actions from happening after component has been unmounted
+  // see https://overreacted.io/a-complete-guide-to-useeffect/#speaking-of-race-conditions
+  useEffect(() => {
+    let didCancel = false;
+
+    const onSubmit = async () => {
```

Comment:
IMHO this can be treated in a callback, `useEffect` hook doesn't seem to me as the right place for handling onClick events but for changing UI state
```js
const onSubmit = async () => {
  setSubmitting(true);
  const response = await onEdit(inputValue, attribute);
 // resposneHandlerForErrors(response)
  setSubmitting(false);
  setEditing(false);
}
          <Button
            aria-label={`submit ${attribute}`}
            variant="plain"
            onClick={() => onSubmit}
          >
```

We can also use redux here with the API middleware behind the scenes, so we can gain  errors handlers out of the box
```js

const onSubmit = () => {
  dispatch(onEdit(url, inputValue, attribute)); //onEdit can be a wrapper for 'post'
}
useSelector((state, attribute) => selectAPIErrorMessage(state, attribute)) // Getting errors messages out of the box
z
```

Reply:
> We can also use redux here with the API middleware behind the scenes, so we can gain errors handlers out of the box

I feel this approach is too tightly coupled to the API middleware, making an assumption that the edit will trigger an API call. I think `onEdit` should handle this. Here is the example from Katello:

- [onEdit is created here](https://github.com/Katello/katello/blob/64f9d6a7dfcf4347ec8937ecf42642adfbed8512/webpack/scenes/ContentViews/Details/ContentViewInfo.js#L46) and passed to `ActionableDetail`, which uses `EditableTextInput` (I changed the name here to `Editable text display`)
- The [redux action](https://github.com/Katello/katello/blob/64f9d6a7dfcf4347ec8937ecf42642adfbed8512/webpack/scenes/ContentViews/Details/ContentViewDetailActions.js#L30-L42) runs and handles errors

Reply:
> IMHO this can be treated in a callback, `useEffect` hook doesn't seem to me as the right place for handling onClick events but for changing UI state

I did have this approach originally but I ran into the race condition issue in testing, so had to add the hook wrapper. Technically a race condition something we should consider but it's unlikely to happen with normal user behavior. It was seen in the functional tests most likely because of the speed in which they execute.

Reply:
It does seem weird to try to track a boolean state change (`submitting`) to decide when to run `onSubmit`.  Can you show me what the code looked like when you used the more traditional approach, and repro steps to see the race condition?

I'm thinking/hoping there must be an alternate way to avoid the race condition that doesn't require this `useEffect` approach.

Reply:
I tried to reproduce again and either it's gone from an update elsewhere or when something else changed in this PR. I switched back to the simple function :+1: